File: .skills/openclaw-skills/skills/0xnyk/xint/lib/sentiment.ts
/**
 * lib/sentiment.ts — AI-powered sentiment analysis via Grok.
 *
 * Enriches tweets with sentiment labels (positive/negative/neutral),
 * confidence scores, and aggregate statistics.
 */

import type { Tweet } from "./api";
import { grokChat, type GrokOpts } from "./grok";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface SentimentResult {
  id: string;
  sentiment: "positive" | "negative" | "neutral" | "mixed";
  score: number;     // -1.0 to 1.0
  label?: string;    // short reason
}

export interface SentimentStats {
  positive: number;
  negative: number;
  neutral: number;
  mixed: number;
  average_score: number;
}

export interface EnrichedTweet extends Tweet {
  sentiment?: SentimentResult;
}

// ---------------------------------------------------------------------------
// Grok sentiment analysis
// ---------------------------------------------------------------------------

const SENTIMENT_SYSTEM = `You are a sentiment analysis engine. Given tweets, return a JSON array with sentiment analysis for each tweet.

For each tweet, provide:
- id: the tweet ID
- sentiment: one of "positive", "negative", "neutral", "mixed"
- score: a number from -1.0 (very negative) to 1.0 (very positive)
- label: a 2-5 word reason (e.g., "excited about launch", "frustrated with bugs", "neutral observation")

Detect sarcasm, irony, and context. Crypto/tech enthusiasm is positive. Complaints/frustration are negative. Questions and factual statements are neutral.

Return ONLY valid JSON array, no markdown fences, no explanation.`;

/**
 * Analyze sentiment for a batch of tweets using Grok.
 * Processes in batches of 20 to stay within token limits.
 */
export async function analyzeSentiment(
  tweets: Tweet[],
  opts?: GrokOpts,
): Promise<SentimentResult[]> {
  if (tweets.length === 0) return [];

  const batchSize = 20;
  const results: SentimentResult[] = [];

  for (let i = 0; i < tweets.length; i += batchSize) {
    const batch = tweets.slice(i, i + batchSize);
    const tweetContext = batch
      .map(t => `[${t.id}] @${t.username}: ${t.text.slice(0, 280)}`)
      .join("\n\n");

    try {
      const response = await grokChat(
        [
          { role: "system", content: SENTIMENT_SYSTEM },
          {
            role: "user",
            content: `Analyze sentiment for these ${batch.length} tweets:\n\n${tweetContext}`,
          },
        ],
        {
          model: opts?.model || "grok-4-1-fast",
          temperature: 0.3,  // low temperature for consistent analysis
          maxTokens: opts?.maxTokens || 2048,
        },
      );

      // Parse JSON response
      const parsed = parseJsonResponse(response.content, batch);
      results.push(...parsed);
    } catch (e: any) {
      // On failure, mark batch as neutral
      console.error(`[sentiment] Batch analysis failed: ${e.message}`);
      for (const t of batch) {
        results.push({ id: t.id, sentiment: "neutral", score: 0, label: "analysis failed" });
      }
    }
  }

  return results;
}

/**
 * Parse Grok's JSON response, with fallback handling for malformed output.
 */
function parseJsonResponse(content: string, tweets: Tweet[]): SentimentResult[] {
  // Strip markdown code fences if present
  let cleaned = content.trim();
  if (cleaned.startsWith("```")) {
    cleaned = cleaned.replace(/^```(?:json)?\n?/, "").replace(/\n?```$/, "");
  }

  try {
    const parsed = JSON.parse(cleaned);
    if (Array.isArray(parsed)) {
      return parsed.map(item => ({
        id: String(item.id),
        sentiment: validateSentiment(item.sentiment),
        score: clampScore(Number(item.score) || 0),
        label: item.label || undefined,
      }));
    }
  } catch {}

  // Fallback: try to extract JSON array from response
  const arrayMatch = cleaned.match(/\[[\s\S]*\]/);
  if (arrayMatch) {
    try {
      const parsed = JSON.parse(arrayMatch[0]);
      if (Array.isArray(parsed)) {
        return parsed.map(item => ({
          id: String(item.id),
          sentiment: validateSentiment(item.sentiment),
          score: clampScore(Number(item.score) || 0),
          label: item.label || undefined,
        }));
      }
    } catch {}
  }

  // Final fallback: return neutral for all
  return tweets.map(t => ({
    id: t.id,
    sentiment: "neutral" as const,
    score: 0,
    label: "parse failed",
  }));
}

function validateSentiment(s: any): SentimentResult["sentiment"] {
  if (["positive", "negative", "neutral", "mixed"].includes(s)) return s;
  return "neutral";
}

function clampScore(n: number): number {
  return Math.max(-1, Math.min(1, Math.round(n * 100) / 100));
}

// ---------------------------------------------------------------------------
// Enrich tweets with sentiment
// ---------------------------------------------------------------------------

/**
 * Attach sentiment results to tweets.
 */
export function enrichTweets(tweets: Tweet[], sentiments: SentimentResult[]): EnrichedTweet[] {
  const sentimentMap = new Map(sentiments.map(s => [s.id, s]));
  return tweets.map(t => ({
    ...t,
    sentiment: sentimentMap.get(t.id),
  }));
}

/**
 * Compute aggregate sentiment stats.
 */
export function computeStats(sentiments: SentimentResult[]): SentimentStats {
  const counts = { positive: 0, negative: 0, neutral: 0, mixed: 0 };
  let totalScore = 0;

  for (const s of sentiments) {
    counts[s.sentiment]++;
    totalScore += s.score;
  }

  return {
    ...counts,
    average_score: sentiments.length > 0
      ? Math.round((totalScore / sentiments.length) * 100) / 100
      : 0,
  };
}

// ---------------------------------------------------------------------------
// Formatting
// ---------------------------------------------------------------------------

function sentimentIcon(s: SentimentResult["sentiment"]): string {
  switch (s) {
    case "positive": return "+";
    case "negative": return "-";
    case "mixed": return "~";
    default: return "=";
  }
}

function scoreBar(score: number): string {
  // Convert -1..1 to 0..10
  const normalized = Math.round((score + 1) * 5);
  const filled = Math.max(0, Math.min(10, normalized));
  return "[" + "#".repeat(filled) + ".".repeat(10 - filled) + "]";
}

/**
 * Format a sentiment-enriched tweet for terminal display.
 */
export function formatSentimentTweet(t: EnrichedTweet, index?: number): string {
  const prefix = index !== undefined ? `${index + 1}. ` : "";
  const s = t.sentiment;
  const sentimentTag = s
    ? `(${sentimentIcon(s.sentiment)}${s.score.toFixed(1)} ${s.label || s.sentiment})`
    : "";

  const cleanText = t.text.replace(/https:\/\/t\.co\/\S+/g, "").trim();
  const truncated = cleanText.length > 180 ? cleanText.slice(0, 177) + "..." : cleanText;

  let out = `${prefix}@${t.username} ${sentimentTag}`;
  out += `\n${truncated}`;
  out += `\n${t.tweet_url}`;

  return out;
}

/**
 * Format aggregate sentiment stats.
 */
export function formatStats(stats: SentimentStats, total: number): string {
  const pct = (n: number) => total > 0 ? `${Math.round((n / total) * 100)}%` : "0%";

  let out = `\nSentiment Analysis (${total} tweets):\n`;
  out += `  Avg score: ${stats.average_score.toFixed(2)} ${scoreBar(stats.average_score)}\n`;
  out += `  + Positive: ${stats.positive} (${pct(stats.positive)})\n`;
  out += `  - Negative: ${stats.negative} (${pct(stats.negative)})\n`;
  out += `  = Neutral:  ${stats.neutral} (${pct(stats.neutral)})\n`;
  if (stats.mixed > 0) {
    out += `  ~ Mixed:    ${stats.mixed} (${pct(stats.mixed)})\n`;
  }

  return out;
}

File: .skills/openclaw-skills/skills/0xnyk/xint-rs/src/sentiment.rs
use crate::api::grok;
use crate::models::*;
use anyhow::Result;
use std::path::Path;

const SENTIMENT_SYSTEM: &str = r#"You are a sentiment analysis engine. Given tweets, return a JSON array with sentiment analysis for each tweet.

For each tweet, provide:
- id: the tweet ID
- sentiment: one of "positive", "negative", "neutral", "mixed"
- score: a number from -1.0 (very negative) to 1.0 (very positive)
- label: a 2-5 word reason (e.g., "excited about launch", "frustrated with bugs", "neutral observation")

Detect sarcasm, irony, and context. Crypto/tech enthusiasm is positive. Complaints/frustration are negative. Questions and factual statements are neutral.

Return ONLY valid JSON array, no markdown fences, no explanation."#;

/// Analyze sentiment for a batch of tweets using Grok.
/// Processes in batches of 20 to stay within token limits.
pub async fn analyze_sentiment(
    http: &reqwest::Client,
    api_key: &str,
    tweets: &[Tweet],
    model: Option<&str>,
    costs_path: Option<&Path>,
) -> Result<Vec<SentimentResult>> {
    if tweets.is_empty() {
        return Ok(Vec::new());
    }

    let batch_size = 20;
    let mut results = Vec::new();

    for chunk in tweets.chunks(batch_size) {
        let tweet_context: String = chunk
            .iter()
            .map(|t| {
                let text = if t.text.len() > 280 {
                    &t.text[..280]
                } else {
                    &t.text
                };
                format!("[{}] @{}: {}", t.id, t.username, text)
            })
            .collect::<Vec<_>>()
            .join("\n\n");

        let messages = vec![
            GrokMessage {
                role: "system".to_string(),
                content: SENTIMENT_SYSTEM.to_string(),
            },
            GrokMessage {
                role: "user".to_string(),
                content: format!(
                    "Analyze sentiment for these {} tweets:\n\n{}",
                    chunk.len(),
                    tweet_context
                ),
            },
        ];

        let opts = GrokOpts {
            model: model.unwrap_or("grok-4-1-fast").to_string(),
            temperature: 0.3,
            max_tokens: 2048,
        };

        match grok::grok_chat_tracked(http, api_key, &messages, &opts, costs_path).await {
            Ok(response) => {
                let parsed = parse_json_response(&response.content, chunk);
                results.extend(parsed);
            }
            Err(e) => {
                eprintln!("[sentiment] Batch analysis failed: {e}");
                for t in chunk {
                    results.push(SentimentResult {
                        id: t.id.clone(),
                        sentiment: "neutral".to_string(),
                        score: 0.0,
                        label: Some("analysis failed".to_string()),
                    });
                }
            }
        }
    }

    Ok(results)
}

/// Parse Grok's JSON response, with fallback handling for malformed output.
fn parse_json_response(content: &str, tweets: &[Tweet]) -> Vec<SentimentResult> {
    let mut cleaned = content.trim().to_string();

    // Strip markdown code fences if present
    if cleaned.starts_with("```") {
        if let Some(start) = cleaned.find('\n') {
            cleaned = cleaned[start + 1..].to_string();
        }
        if cleaned.ends_with("```") {
            cleaned = cleaned[..cleaned.len() - 3].trim().to_string();
        }
    }

    // Try parsing as JSON array
    if let Ok(arr) = serde_json::from_str::<Vec<serde_json::Value>>(&cleaned) {
        return arr.iter().map(parse_sentiment_item).collect();
    }

    // Fallback: try to extract JSON array from response
    if let Some(start) = cleaned.find('[') {
        if let Some(end) = cleaned.rfind(']') {
            let slice = &cleaned[start..=end];
            if let Ok(arr) = serde_json::from_str::<Vec<serde_json::Value>>(slice) {
                return arr.iter().map(parse_sentiment_item).collect();
            }
        }
    }

    // Final fallback: return neutral for all
    tweets
        .iter()
        .map(|t| SentimentResult {
            id: t.id.clone(),
            sentiment: "neutral".to_string(),
            score: 0.0,
            label: Some("parse failed".to_string()),
        })
        .collect()
}

fn parse_sentiment_item(item: &serde_json::Value) -> SentimentResult {
    let id = item
        .get("id")
        .map(|v| match v {
            serde_json::Value::String(s) => s.clone(),
            serde_json::Value::Number(n) => n.to_string(),
            _ => "?".to_string(),
        })
        .unwrap_or_else(|| "?".to_string());

    let sentiment = item
        .get("sentiment")
        .and_then(|v| v.as_str())
        .map(validate_sentiment)
        .unwrap_or("neutral")
        .to_string();

    let score = item
        .get("score")
        .and_then(|v| v.as_f64())
        .map(|s| s.clamp(-1.0, 1.0))
        .map(|s| (s * 100.0).round() / 100.0)
        .unwrap_or(0.0);

    let label = item
        .get("label")
        .and_then(|v| v.as_str())
        .map(|s| s.to_string());

    SentimentResult {
        id,
        sentiment,
        score,
        label,
    }
}

fn validate_sentiment(s: &str) -> &str {
    match s {
        "positive" | "negative" | "neutral" | "mixed" => s,
        _ => "neutral",
    }
}

/// Compute aggregate sentiment stats.
pub fn compute_stats(sentiments: &[SentimentResult]) -> SentimentStats {
    let mut positive = 0u32;
    let mut negative = 0u32;
    let mut neutral = 0u32;
    let mut mixed = 0u32;
    let mut total_score = 0.0f64;

    for s in sentiments {
        match s.sentiment.as_str() {
            "positive" => positive += 1,
            "negative" => negative += 1,
            "mixed" => mixed += 1,
            _ => neutral += 1,
        }
        total_score += s.score;
    }

    let average_score = if sentiments.is_empty() {
        0.0
    } else {
        (total_score / sentiments.len() as f64 * 100.0).round() / 100.0
    };

    SentimentStats {
        positive,
        negative,
        neutral,
        mixed,
        average_score,
    }
}

/// Format aggregate sentiment stats for terminal display.
pub fn format_stats(stats: &SentimentStats, total: usize) -> String {
    let pct = |n: u32| -> String {
        if total > 0 {
            format!("{}%", (n as f64 / total as f64 * 100.0).round() as u32)
        } else {
            "0%".to_string()
        }
    };

    let score_bar = {
        let normalized = ((stats.average_score + 1.0) * 5.0).round() as usize;
        let filled = normalized.min(10);
        format!("[{}{}]", "#".repeat(filled), ".".repeat(10 - filled))
    };

    let mut out = format!("\nSentiment Analysis ({total} tweets):\n");
    out.push_str(&format!(
        "  Avg score: {:.2} {}\n",
        stats.average_score, score_bar
    ));
    out.push_str(&format!(
        "  + Positive: {} ({})\n",
        stats.positive,
        pct(stats.positive)
    ));
    out.push_str(&format!(
        "  - Negative: {} ({})\n",
        stats.negative,
        pct(stats.negative)
    ));
    out.push_str(&format!(
        "  = Neutral:  {} ({})\n",
        stats.neutral,
        pct(stats.neutral)
    ));
    if stats.mixed > 0 {
        out.push_str(&format!(
            "  ~ Mixed:    {} ({})\n",
            stats.mixed,
            pct(stats.mixed)
        ));
    }

    out
}
